src/cdomains/congruenceClosure.ml

```diff
@@ -0,0 +1,1601 @@
+(** OCaml implementation of a quantitative congruence closure. *)
+
+open Batteries
+open GoblintCil
+module Var = CilType.Varinfo
+module M = Messages
+
+exception Unsat
+
+type ('v, 't) term = Addr of 'v | Deref of ('v, 't) term * Z.t * 't [@@deriving eq, ord, hash]
+type ('v, 't) prop = Equal of ('v, 't) term * ('v, 't) term * Z.t | Nequal of ('v, 't) term * ('v, 't) term * Z.t [@@deriving eq, ord, hash]
+
+(** The terms consist of address constants and dereferencing function with sum of an integer.
+    The dereferencing function is parametrized by the size of the element in the memory.
+    We store the CIL expression of the term in the data type, such that it it easier to find the types of the dereferenced elements.
+    Also so that we can easily convert back from term to Cil expression.
+*)
+module T = struct
+  type exp = Cil.exp
+
+  let bitsSizeOfPtr () = Z.of_int @@ bitsSizeOf (TPtr (TVoid [],[]))
+
+  (* equality of terms should not depend on the expression *)
+  let compare_exp _ _ = 0
+  let equal_exp _ _ = true
+  let hash_exp _ = 1
+
+
+  (* we store the varinfo and the Cil expression corresponding to thi term in the data type *)
+  type t = (Var.t, exp) term [@@deriving eq, ord, hash]
+  type v_prop = (Var.t, exp) prop [@@deriving ord, hash]
+
+  let compare t1 t2 =
+    match t1,t2 with
+    | Addr v1, Addr v2 -> Var.compare v1 v2
+    | Deref (t1,z1,_), Deref (t2,z2,_) -> let c = compare t1 t2 in
+      if c = 0 then Z.compare z1 z2 else c
+    | Addr _, Deref _ -> -1
+    | Deref _, Addr _ -> 1
+
+  (** Two propositions are equal if they are syntactically equal
+      or if one is t_1 = z + t_2 and the other t_2 = - z + t_1. *)
+  let equal_v_prop p1 p2 =
+    let equivalent_triple (t1,t2,o1) (t3,t4,o2) =
+      (equal t1 t3 && equal t2 t4 && Z.equal o1 o2) ||
+      (equal t1 t4 && equal t2 t3 && Z.(equal o1 (-o2)))
+    in match p1, p2 with
+    | Equal (a,b,c), Equal (a',b',c') -> equivalent_triple (a,b,c) (a',b',c')
+    | Nequal (a,b,c), Nequal (a',b',c') -> equivalent_triple (a,b,c) (a',b',c')
+    | _ -> false
+
+  let compare_v_prop p1 p2 =
```

Comment:
Does this compare respect the rules for a compare?
Consider p1 and p2 that compare equal according to the `equivalent _triple` rule above, and some p' that compares smaller than p1.
Is it guaranteed to also compare smaller than p2?

If not, many data structures that rely on compare providing an order will silently provide incorrect results or have horrible run-time complexity.